flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java

```diff
@@ -0,0 +1,139 @@
+ * </ul>
+ *
+ * <p>Example:
+ *
+ * <pre>{@code
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00')  // Parses string using default format
+ * TO_TIMESTAMP_LTZ(1234567890123)  // Converts epoch milliseconds
+ * TO_TIMESTAMP_LTZ(1234567890, 0)     // Converts epoch seconds
+ * TO_TIMESTAMP_LTZ(1234567890123, 3)  // Converts epoch milliseconds
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00')  // Parses string using default format
+ * TO_TIMESTAMP_LTZ('2023-01-01T10:00:00', 'yyyy-MM-dd\'T\'HH:mm:ss') // Parses string using input format
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00', 'yyyy-MM-dd HH:mm:ss', 'UTC') // Parses string using input format and timezone
+ * }</pre>
+ */
+@Internal
+public class ToTimestampLtzFunction extends BuiltInScalarFunction {
+
+    private static final int DEFAULT_PRECISION = 3;
+
+    public ToTimestampLtzFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.TO_TIMESTAMP_LTZ, context);
+    }
+
+    public @Nullable TimestampData eval(Number epoch, Integer precision) {
+        if (epoch == null || precision == null) {
+            return null;
+        }
+        if (epoch instanceof Float || epoch instanceof Double) {
+            return DateTimeUtils.toTimestampData(epoch.doubleValue(), precision);
+        }
+        return DateTimeUtils.toTimestampData(epoch.longValue(), precision);
+    }
+
+    public @Nullable TimestampData eval(DecimalData epoch, Integer precision) {
+        if (epoch == null || precision == null) {
+            return null;
+        }
+
+        return DateTimeUtils.toTimestampData(epoch, precision);
+    }
+
+    public @Nullable TimestampData eval(Number epoch) {
+        return eval(epoch, DEFAULT_PRECISION);
+    }
+
+    public @Nullable TimestampData eval(DecimalData epoch) {
+        return eval(epoch, DEFAULT_PRECISION);
+    }
+
+    public @Nullable TimestampData eval(StringData timestamp) {
+        if (timestamp == null) {
+            return null;
+        }
+
+        return parseTimestampData(timestamp.toString());
+    }
+
+    public @Nullable TimestampData eval(StringData timestamp, StringData format) {
+        if (timestamp == null || format == null) {
+            return null;
+        }
+
+        return parseTimestampData(timestamp.toString(), format.toString());
+    }
+
+    public @Nullable TimestampData eval(
+            StringData dateStr, StringData format, StringData timezone) {
+        if (dateStr == null || format == null || timezone == null) {
+            return null;
+        }
+
+        TimestampData ts = parseTimestampData(dateStr.toString(), format.toString());
+        if (ts == null) {
+            return null;
+        }
+
+        try {
+            ZonedDateTime zoneDate = ts.toLocalDateTime().atZone(ZoneId.of(timezone.toString()));
+            return TimestampData.fromInstant(zoneDate.toInstant());
+        } catch (DateTimeException e) {
+            return null;
```

Comment:
LGTM, just a question, why we return null in catch block?

Reply:
Good question - should we log this exception rather than just swallowing it?

Reply:
Good question, @davidradl and @hanyuzheng7 !

I explained the reasoning here: https://github.com/apache/flink/pull/25763#discussion_r1893222879
We return null here because it's the desired behavior.
Also because it's consistent to how we handle exceptions in DateTime, eg. [here](https://github.com/apache/flink/blob/7fa4f78e43bfd99a7e7a3d655a9a3641e4d7bc59/flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java#L447). 

Logging would be fine, but that also requires refactoring other functions to make it consistent, maybe in a separate effort.

Reply:
No logging in the “hot path“ please
We are building a streaming/batch system that processes millions of events per minute
 Logging could fill a local TM disk or the log provider easily.

Reply:
Good to know the specific reasoning! TIL